### PR TITLE
Vendoring vishvananda/netlink @c682914

### LINKED ISSUE #32090: Docker leaks container netns netlink sockets to forked docker-proxy processes, preventing netns cleanup after container is stopped
I have been debugging a networking issue when using Weave networking, where Weave overlay networking addresses associated with Docker containers continue to respond to pings even after the Docker container has been destroyed. This causes issues when the Weave overlay networking address is re-allocated for use by a different container, as both the legit container and the undead zombie container continue to respond got ARP queries for the IP address. Clients are randomly unable to connect to services when their traffic ends up at the zombie container: https://github.com/weaveworks/weave/issues/2842 https://github.com/weaveworks/weave/issues/2433#issuecomment-265993860

This is fundamentally caused by leaking references to the container's netns. When a Docker container stops, Docker will explicitly delete its own `eth0` veth interface pair before unmounting the `/run/docker/netns/*`. For the Weave `ethwe` veth pair, we rely on the interfaces being destroyed by the kernel when the last reference to the container's network namespace goes away. Unfortunately, there are multiple cases where these network namespace references can leak, meaning that the container's network namespace does not get destroyed by the kernel after Docker considers the container to be destroyed. With normal Docker networking, these empty network namespaces are harmless, but when using Weave networking, the containers turn into ghosts that continue to haunt the network :ghost: 

So far, I have found two cases of this happening:

* https://github.com/kontena/kontena/issues/2004 Running cadvisor in a Docker container with the default `rprivate` mode for the `-v /:/rootfs` bind-mount will capture any `/run/docker/netns/*` mounts that exist when the cadvisor container is started, and these may ever not go away when Docker later unlinks the `/run/docker/netns/*` mountpoint, depending on the kernel version
* This issue, where `dockerd` launching a new `docker-proxy` process captures references to any running Docker container network namspaces in the form of netlink sockets inherited from the forking parent

Please refer to my earlier issue comments for further background on this issue:

* https://github.com/weaveworks/weave/issues/2433#issuecomment-265993860
* https://github.com/kontena/kontena/issues/2004
* https://gist.github.com/SpComb/9cc2a7ed46ff708547d09854b78c197f#weave-zombie-huntersh

And the debugging for this specific issue:
* https://github.com/weaveworks/weave/issues/2842#issuecomment-288054662 onwards
* https://gist.github.com/SpComb/f83c53b05aad79c6213185a3da7a7902
* https://gist.github.com/SpComb/197cf7c4191dcca261fdfeade0be3c54

## Description

It seems that the `dockerd` process holds open a netlink socket for each running Docker container, associated with the container's network namespace. It also seems that when Docker forks a new `docker-proxy` process, these netlink socket FDs are inherited by the `docker-proxy` process: https://github.com/weaveworks/weave/issues/2842#issuecomment-289022778

After Docker stops a container and closes its own netlink socket, that container's network namespace will not be cleaned up by the kernel until the `docker-proxy` processes for some unrelated container exit, and these leaked netlink sockets are closed by the kernel: https://github.com/weaveworks/weave/issues/2842#issuecomment-288791826

## Steps to reproduce the issue

See the gist: https://gist.github.com/SpComb/c8604792fc9d3ba0518d889ff8f89ccb

1. `git clone https://gist.github.com/SpComb/c8604792fc9d3ba0518d889ff8f89ccb docker-netns-leak && cd docker-netns-leak`
2. `sudo ./test.sh`

Run a `victim` container, and attach an extra veth pair connected to a `br-test` bridge, configured with a `192.168.0.10` address.

Run a second `bystander` container, using `-p 6379:6379` to publish a port. A new `docker-proxy` process is forked, which inherits the netlink socket from the `victim` network namespace.

The `victim` container is stopped, and the extra veth pair does not get destroyed, and the `192.168.0.10` address inside the stopped container continues to respond to pings.

This does not happen unless the `docker-proxy` process is started while the `victim` container is running, and the extra veth pair from the `victim` container goes away once the `docker-proxy` process is stopped.

## Describe the results you received

Stopping the `victim` container only destroys the Docker veth pair:

```
80: veth102a2ef@veth84ba9ce: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
81: veth84ba9ce@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue master docker0 state DOWN group default 
    link/ether 6e:25:9f:c2:3c:9e brd ff:ff:ff:ff:ff:ff
Deleted 80: veth102a2ef@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 81: veth84ba9ce@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether 6e:25:9f:c2:3c:9e
```

The extra `veth` pair attached to the stopped `victim` container's network namespace does not get destroyed, while the `bystander` container is still running with its `docker-proxy` process:

```
83: veth0@if82: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-test state UP mode DEFAULT group default qlen 1000
    link/ether b2:3c:93:87:53:20 brd ff:ff:ff:ff:ff:ff
85: veth8897a0a@if84: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP mode DEFAULT group default 
    link/ether c6:f9:27:97:fb:b3 brd ff:ff:ff:ff:ff:ff

```

It also continues to responds to pings:

```
PING 192.168.0.10 (192.168.0.10) 56(84) bytes of data.
64 bytes from 192.168.0.10: icmp_seq=1 ttl=64 time=0.118 ms
64 bytes from 192.168.0.10: icmp_seq=2 ttl=64 time=0.148 ms
```

Stopping the `bystander` container will then stop the `docker-proxy` process, and the `victim` container's extra veth pair gets destroyed, as well as the `bystander` container's Docker veth pair:

```
Deleted 83: veth0@NONE: <BROADCAST,MULTICAST> mtu 1500 master br-test state DOWN 
    link/ether b2:3c:93:87:53:20

84: vethbf7273c@veth8897a0a: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
Deleted 84: vethbf7273c@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
Deleted 85: veth8897a0a@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether c6:f9:27:97:fb:b3
```

## Describe the results you expected

If the `bystander` container with its associated `docker-proxy` is not running, then stopping the `victim` container will correctly destroy the Docker veths, release the network namespace and thus also cause the extra veths to be destroyed immediately:

```
76: veth91821c3@veth0b65d5c: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 76: veth91821c3@NONE: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
Deleted 77: veth0b65d5c@NONE: <BROADCAST,MULTICAST> mtu 1500 master docker0 state DOWN 
    link/ether 4e:c3:f5:87:c8:dd

Deleted 79: veth0@NONE: <BROADCAST,MULTICAST> mtu 1500 master br-test state DOWN 
    link/ether 8e:44:24:a5:c3:67
```

## Output of `docker version`:

```
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue Mar 14 21:10:49 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5236f0
 Built:        Tue Mar 14 21:10:49 2017
 OS/Arch:      linux/amd64
```

## Output of `docker info`:

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.12.6
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp selinux
Kernel Version: 4.9.9-coreos-r1
Operating System: Container Linux by CoreOS 1298.6.0 (Ladybug)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.9 MiB
Name: coreos-1gb-ams2-01
ID: HLW2:MFQE:MQHF:UWLC:R6VH:E2HW:7GG2:WEMD:DZSI:GSGS:A5HA:HK5E
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

## Additional environment details

CoreOS stable 1298.6.0 on DigitalOcean.

### PR DESCRIPTION
**Pull request by aboch**
Fixes #32090

### CODE REVIEW

File: vendor/github.com/vishvananda/netlink/nl/nl_linux.go [hunk 1]
```diff
@@ -450,7 +451,7 @@ type NetlinkSocket struct {
 }
 
 func getNetlinkSocket(protocol int) (*NetlinkSocket, error) {
-	fd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, protocol)
+	fd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW|syscall.SOCK_CLOEXEC, protocol)
```
Comment (aboch) [inline vendor/github.com/vishvananda/netlink/nl/nl_linux.go]: This is the relevant change

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by tophj-ibm** [comment]
@aboch this is probably fixed by your libnetwork patch right? Looks similar to #32173

```sh
13:56:29 ----------------------------------------------------------------------
13:56:29 FAIL: check_test.go:353: DockerSwarmSuite.TearDownTest
13:56:29 
13:56:29 check_test.go:358:
13:56:29     d.Stop(c)
13:56:29 daemon/daemon.go:392:
13:56:29     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
13:56:29 ... Error: Error while stopping the daemon d33e0b21cea68 : exit status 2
13:56:29 
13:56:29 
13:56:29 ----------------------------------------------------------------------
13:56:29 PANIC: docker_cli_swarm_test.go:418: DockerSwarmSuite.TestSwarmIngressNetwork
13:56:29 
13:56:29 [d33e0b21cea68] waiting for daemon to start
13:56:29 [d33e0b21cea68] daemon started
13:56:29 
13:56:29 [d33e0b21cea68] exiting daemon
13:56:29 ... Panic: Fixture has panicked (see related PANIC)
13:56:29 
13:56:29 ----------------------------------------------------------------------
```

**Comment by aboch** [comment]
Thanks @tophj-ibm 
Let's check the CI damon logs to confirm

**Comment by tophj-ibm** [comment]
@aboch pretty it should be fixed by your patch. This look about right?
```sh
time="2017-03-29T18:55:11.646405000Z" level=error msg="Failed to delete ingress sandbox: another sandbox delete in progress" 
time="2017-03-29T18:55:11.771623000Z" level=debug msg="checkEncryption(vjzkr6n, <nil>, 0, true)" 
time="2017-03-29T18:55:11.771747000Z" level=debug msg="checkEncryption(vjzkr6n, <nil>, 0, true)" 
time="2017-03-29T18:55:11.811836000Z" level=debug msg="Remove interface veth3c0bec4 failed: Link not found" 
panic: runtime error: index out of range
```

**Comment by aboch** [comment]
@tophj-ibm 
Yes it's the same:

```
time="2017-03-29T18:55:11.811836000Z" level=debug msg="Remove interface veth3c0bec4 failed: Link not found" 
panic: runtime error: index out of range

goroutine 198 [running]:
panic(0x113a93a0, 0xc420014090)
	/usr/local/go/src/runtime/panic.go:500 +0x3a8
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*epHeap).Swap(0xc4209ef450, 0xffffffffffffffff, 0x0)
	<autogenerated>:69 +0xd8
container/heap.Remove(0x11f363c0, 0xc4209ef450, 0xffffffffffffffff, 0x0, 0x0)
	/usr/local/go/src/container/heap/heap.go:74 +0x80
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*sandbox).clearNetworkResources(0xc4209ef2c0, 0xc4214982c0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/sandbox.go:920 +0x440
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*endpoint).sbLeave(0xc4214982c0, 0xc4209ef2c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/endpoint.go:703 +0x990
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*endpoint).Delete(0xc421498160, 0xc4214ea001, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/endpoint.go:781 +0x414
github.com/docker/docker/daemon.(*Daemon).releaseIngress(0xc4203fc200, 0xc4213ae3c0, 0x19)
	/go/src/github.com/docker/docker/daemon/network.go:220 +0x2e0
github.com/docker/docker/daemon.(*Daemon).startIngressWorker.func1(0xc4203fc200)
	/go/src/github.com/docker/docker/daemon/network.go:124 +0x13c
created by github.com/docker/docker/daemon.(*Daemon).startIngressWorker
	/go/src/github.com/docker/docker/daemon/network.go:129 +0x74
```

**Comment by tophj-ibm** [comment]
@aboch thanks for confirming and fixing it! I'll restart that job

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**